web/src/queries/zfcp.ts

```diff
@@ -0,0 +1,193 @@
+  queryFn: supportedZFCP,
+};
+
+const zfcpConfigQuery = {
+  queryKey: ["zfcp", "config"],
+  queryFn: fetchZFCPConfig,
+};
+
+/**
+ * Hook that returns zFCP controllers.
+ */
+const useZFCPControllers = (): ZFCPController[] => {
+  const { data: controllers } = useSuspenseQuery(zfcpControllersQuery);
+  return controllers;
+};
+
+/**
+ * Hook that returns zFCP disks.
+ */
+const useZFCPDisks = (): ZFCPDisk[] => {
+  const { data: devices } = useSuspenseQuery(zfcpDisksQuery);
+  return devices;
+};
+
+/**
+ * Hook that returns zFCP config.
+ */
+const useZFCPSupported = (): boolean => {
+  const { data: supported } = useSuspenseQuery(zfcpSupportedQuery);
+  return supported;
+};
+/**
+ * Hook that returns zFCP config.
+ */
+const useZFCPConfig = (): ZFCPConfig => {
+  const { data: config } = useSuspenseQuery(zfcpConfigQuery);
+  return config;
+};
+
+/**
+ * Listens for zFCP Controller changes.
+ */
+const useZFCPControllersChanges = () => {
+  const client = useInstallerClient();
+  const queryClient = useQueryClient();
+
+  React.useEffect(() => {
+    if (!client) return;
+
+    return client.ws().onEvent((event) => {
+      switch (event.type) {
+        case "ZFCPControllerAdded": {
+          const device: ZFCPController = event.device;
+          queryClient.setQueryData(
+            zfcpControllersQuery.queryKey,
+            (prev: ZFCPController[] | undefined) => {
+              if (prev === undefined) return;
+              return [...prev, device];
+            },
+          );
+          break;
+        }
+        case "ZFCPControllerRemoved": {
+          const device: ZFCPController = event.device;
+          const { id } = device;
+          queryClient.setQueryData(
+            zfcpControllersQuery.queryKey,
+            (prev: ZFCPController[] | undefined) => {
+              if (prev === undefined) return;
+              const res = prev.filter((dev) => dev.id !== id);
+              return res;
+            },
+          );
+          break;
+        }
+        case "ZFCPControllerChanged": {
+          const device: ZFCPController = event.device;
+          const { id } = device;
+          queryClient.setQueryData(
+            zfcpControllersQuery.queryKey,
+            (prev: ZFCPController[] | undefined) => {
+              if (prev === undefined) return;
+              // deep copy of original to have it immutable
+              const res = [...prev];
+              const index = res.findIndex((dev) => dev.id === id);
+              res[index] = device;
+              return res;
+            },
+          );
+          break;
+        }
+      }
+    });
+  });
+};
+
+/**
+ * Listens for zFCP disks changes.
+ */
+const useZFCPDisksChanges = () => {
+  const client = useInstallerClient();
+  const queryClient = useQueryClient();
+
+  React.useEffect(() => {
+    if (!client) return;
+
+    return client.ws().onEvent((event) => {
+      switch (event.type) {
+        case "ZFCPDiskAdded": {
```

Comment:
This code looks too complex to me. You are calling `setQueryData` three times, passing different functions. I would do it the other way around. Please, see [the iSCSI example](https://github.com/openSUSE/agama/blob/bc057f474f9783717f355af970440e7215b2b705/web/src/queries/storage/iscsi.ts#L109).